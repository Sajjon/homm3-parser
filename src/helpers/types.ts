export type FlaggedProp<FlagName extends string, ValueName extends string, T> =
  | Record<FlagName, true> & Record<ValueName, T>
  | Record<FlagName, false>
