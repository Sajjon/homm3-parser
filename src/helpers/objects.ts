import { ObjectType, objectTypeToMeta } from '../parsers/h3m/enums/object'
import { MetaType } from '../parsers/h3m/contracts/enums/MetaType'
import { createAdapter, TagProducer, TagOrWrapper } from 'binary-markup'

export function getMetaType(type: ObjectType): MetaType {
  if (type in objectTypeToMeta) {
    return objectTypeToMeta[type]
  }
  return MetaType.None
}

export const bitMasksArray = <T>(
  tag: TagOrWrapper<number[]>,
  bitGroups: [number, T][][],
): TagProducer<T[]> =>
  createAdapter<number[], T[]>(
    tag,
    (data: number[]) => {
      return data.reduce((acc: T[], mask, index) => {
        const bits = bitGroups[index]
        for (const [bit, skill] of bits) {
          if (mask & bit) {
            acc.push(skill)
          }
        }
        return acc
      }, [])
    },
    (skills: T[]) => {
      return bitGroups.map(bits => {
        let mask = 0
        for (const [bit, skill] of bits) {
          if (skills.includes(skill)) {
            mask |= bit
          }
        }
        return mask
      })
    },
  )
