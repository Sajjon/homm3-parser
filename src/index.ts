import { parse, StreamInput, TagOrWrapper } from 'binary-markup'
import { inflate } from 'pako'
import { h3mFile, H3MFile } from './parsers/h3m/h3m.bml'

export * from './parsers/assets'

export { H3MFile } from './parsers/h3m/h3m.bml'

export function parseH3M(data: StreamInput): H3MFile {
  return parse<H3MFile>(h3mFile, data)
}

export function parseH3MFile(data: Uint8Array): H3MFile {
  const content = inflate(data)
  return parseH3M(content)
}

export function parseAsset<T>(type: TagOrWrapper<T>, data: StreamInput): T {
  return parse(type, data)
}
