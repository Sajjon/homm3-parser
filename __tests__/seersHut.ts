import { testMap } from './helpers/utils'
import { H3MFile } from '../src'

describe('seers hut', () => {
  testMap('seers hut', 'seers-hut.h3m', (map: H3MFile) => map.objectDetails)
})
