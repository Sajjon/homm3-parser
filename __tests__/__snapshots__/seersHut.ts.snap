// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`seers hut should parse seers hut 1`] = `
Object {
  "count": 14,
  "entries": Array [
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": 63,
          "messages": Object {
            "completionMessage": "I thought you had promise.  You have indeed reached level 5. Come in, come in.  Here, I have something to reward you for your efforts.  Do you accept?",
            "progressMessage": "Faugh.  You again.  Come back when you are level 5, as I told you.",
            "proposalMessage": "I am old and wise, and I do not admit just anyone into my home.  You may enter when you have reached experience level 5.",
          },
          "objective": 5,
          "questType": "Experience",
        },
        "reward": 1000,
        "rewardType": "Experience",
      },
      "oaIndex": 2,
      "x": 2,
      "y": 10,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "I've always wanted to meet someone as famous as you.  Will you let me write down your life story?",
            "progressMessage": "Have you found a great hero for me to interview?  He must have reached Attack Skill 1, Defense Skill 2, Spell Power 3 and Knowledge 4.",
            "proposalMessage": "I am a biographer of great heroes.  I'd really like to meet a hero who has mastered Attack Skill 1, Defense Skill 2, Spell Power 3 and Knowledge 4.  I'd pay well for his story.",
          },
          "objective": Object {
            "attack": 1,
            "defense": 2,
            "knowledge": 4,
            "spellPower": 3,
          },
          "questType": "PrimarySkills",
        },
        "reward": 10,
        "rewardType": "SpellPoints",
      },
      "oaIndex": 2,
      "x": 5,
      "y": 10,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "Ah, exactly what I needed!  Here is the reward I promised.  You still wish to trade the Angel Wings and Arms of Legion, yes?",
            "progressMessage": "Nothing, eh?  I'm sure you will find the Angel Wings and Arms of Legion soon.  Please keep looking.",
            "proposalMessage": "Long ago, powerful wizards were able to create magical artifacts, but time has caused us to forget how to make new items.  I would like to learn these techniques myself, but I need one of these artifacts first to see how it was done.  If you could bring me, the Angel Wings and Arms of Legion, you would be well rewarded. ",
          },
          "objective": Object {
            "artifacts": Array [
              "AngelWings",
              "ArmsOfLegion",
            ],
            "count": 2,
          },
          "questType": "Artifacts",
        },
        "reward": Object {
          "quantity": 10,
          "resource": "Sulfur",
        },
        "rewardType": "Resource",
      },
      "oaIndex": 2,
      "x": 2,
      "y": 12,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "At last, the 1 Archer and 10 Monks that will save our empire!  Here is your payment. Are they ready to depart?",
            "progressMessage": "No luck in finding the 1 Archer and 10 Monks?  Please hurry, the empire depends on you.",
            "proposalMessage": "I am an agent for an emperor of a distant land.  Recently, his armies have fallen on hard times.  If you could bring 1 Archer and 10 Monks to me, I could pay you handsomely.",
          },
          "objective": Object {
            "count": 2,
            "creatures": Array [
              Object {
                "quantity": 1,
                "type": "Archer",
              },
              Object {
                "quantity": 10,
                "type": "Monk",
              },
            ],
          },
          "questType": "Creatures",
        },
        "reward": Object {
          "bonus": 2,
          "skill": "Defence",
        },
        "rewardType": "PrimarySkill",
      },
      "oaIndex": 2,
      "x": 5,
      "y": 12,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "I thought the day would never come!  Sorsha is no more.  Please, will you accept this reward?",
            "progressMessage": "Oh, I wish you brought better news.  It aches my heart that Sorsha still roams free.",
            "proposalMessage": "I was once rich and famous, but Sorsha the terrible was my downfall.  I lost my lands, I lost my title, and I lost my family.  Please, bring the villain to justice.",
          },
          "objective": 372973065,
          "questType": "DefeatHero",
        },
        "reward": 3,
        "rewardType": "Morale",
      },
      "oaIndex": 2,
      "x": 8,
      "y": 10,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 255,
        "header": "ah00_e.def",
        "landscapeGroup": 255,
        "metaType": "Hero",
        "objectGroup": "Heroes",
        "objectSubType": 0,
        "objectType": "Hero",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "absodId": 372973065,
        "artifacts": null,
        "biography": Object {
          "biography": null,
          "hasBiography": false,
        },
        "creatures": null,
        "experience": null,
        "face": null,
        "formation": "Spread",
        "gender": "Default",
        "hasArtifacts": false,
        "hasCreatures": false,
        "hasExperience": false,
        "hasFace": false,
        "hasName": false,
        "hasPrimarySkills": false,
        "hasSecondarySkills": false,
        "hasSpells": false,
        "hero": "Sorsha",
        "name": null,
        "owner": "Red",
        "patrolRadius": 255,
        "primarySkills": null,
        "secondarySkills": null,
        "spells": null,
      },
      "oaIndex": 3,
      "x": 11,
      "y": 10,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "At last, you defeated the Peasants, and the countryside is safe again!  Are you ready to accept the reward?",
            "progressMessage": "Don't lose heart.  Defeating the Peasants is a difficult task, but you will surely succeed.",
            "proposalMessage": "Peasants are menacing the northern region of this land.  If you could be so bold as to defeat them, I would reward you richly.",
          },
          "objective": 372973066,
          "questType": "DefeatMonster",
        },
        "reward": 2,
        "rewardType": "Luck",
      },
      "oaIndex": 2,
      "x": 13,
      "y": 10,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVWpeas.def",
        "landscapeGroup": 255,
        "metaType": "Creature",
        "objectGroup": "Creatures",
        "objectSubType": "Peasant",
        "objectType": "Creature",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "absodId": 372973066,
        "disposition": "Aggressive",
        "doesNotGrow": false,
        "hasMessageAndTreasure": false,
        "messageAndTreasure": null,
        "neverFlees": false,
        "quantity": 0,
      },
      "oaIndex": 4,
      "x": 15,
      "y": 10,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "Finally!  Here, give the 1 Wood, 2 Mercury, 3 Ore, 4 Sulfur, 5 Crystal, 6 Gems and 7 Gold to me, and I'll give you this in return.",
            "progressMessage": "Oh my, that's simply not enough.  I need 1 Wood, 2 Mercury, 3 Ore, 4 Sulfur, 5 Crystal, 6 Gems and 7 Gold.  I'll never complete it with what you have.",
            "proposalMessage": "I am researching a way to turn base metals into gold, but I am short of materials for my workshop.  If you could bring me 1 Wood, 2 Mercury, 3 Ore, 4 Sulfur, 5 Crystal, 6 Gems and 7 Gold, I would be most grateful.",
          },
          "objective": Array [
            1,
            2,
            3,
            4,
            5,
            6,
            7,
          ],
          "questType": "Resources",
        },
        "reward": Object {
          "level": "Advanced",
          "type": "Navigation",
        },
        "rewardType": "SecondarySkill",
      },
      "oaIndex": 2,
      "x": 8,
      "y": 12,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "Finally!  It is you, Malcom.  Here is what I have for you.  Do you accept?",
            "progressMessage": "You are not Malcom.  Begone!",
            "proposalMessage": "What I have is for Malcom alone.  I shall give it to none other.",
          },
          "objective": "Malcom",
          "questType": "BeHero",
        },
        "reward": "AngelicAlliance",
        "rewardType": "Artifact",
      },
      "oaIndex": 2,
      "x": 13,
      "y": 12,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "Ah, one who bears the tan flag.  Here is a prize for you.  Do you accept?",
            "progressMessage": "Your flag is not tan.  I have nothing for you.  Begone!",
            "proposalMessage": "I have a prize for those who fly the tan flag.",
          },
          "objective": "Tan",
          "questType": "BePlayer",
        },
        "reward": "MagicMirror",
        "rewardType": "Spell",
      },
      "oaIndex": 2,
      "x": 2,
      "y": 14,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": -1,
          "messages": Object {
            "completionMessage": "",
            "progressMessage": "",
            "proposalMessage": "",
          },
          "objective": "Red",
          "questType": "BePlayer",
        },
        "reward": Object {
          "quantity": 2,
          "type": "Crusader",
        },
        "rewardType": "Creature",
      },
      "oaIndex": 2,
      "x": 2,
      "y": 16,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseeb0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 2,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": null,
          "messages": null,
          "objective": null,
          "questType": "NONE",
        },
        "reward": null,
        "rewardType": "NONE",
      },
      "oaIndex": 5,
      "x": 3,
      "y": 20,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXseey0.def",
        "landscapeGroup": 255,
        "metaType": "SeersHut",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "SeerHut",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          191,
        ],
      },
      "body": Object {
        "artifactType": null,
        "quest": Object {
          "deadline": null,
          "messages": null,
          "objective": null,
          "questType": "NONE",
        },
        "reward": null,
        "rewardType": "NONE",
      },
      "oaIndex": 2,
      "x": 6,
      "y": 20,
      "z": 0,
    },
  ],
}
`;
